### ISSUE TITLE
Provide a basic GUI to the anonymisation capability

### ISSUE THREAD

**Issue by sjswerdloff** [pymedphys/pymedphys#1022]
Have a simple front end GUI to either the API or the CLI
Support for the pseudonymisation capability would be nice as part of this.

**Comment by SimonBiggs** [comment]
I have been developing a picture of what would be an ideal combination to achieve this in a way that facilitates https://github.com/pymedphys/pymedphys/issues/830. I would want this GUI to be able to be downloaded and installed, have it not depend on a user install of Python (instead have a specific set of tested and pinned libraries). I don't want to use pyinstaller, as we don't really need to shrink down the download size, instead I want to use official embeddable python. And to add new sections to the application, I would like it to be simple for fresh developers.

So I have been narrowing down on the following technologies:

* https://github.com/samuelhwilliams/Eel/
* https://docs.python.org/3/using/windows.html#windows-embeddable
* https://nsis.sourceforge.io/Main_Page
* https://www.streamlit.io/

Let me know what you think.

Cheers,
Simon

**Comment by sjswerdloff** [comment]
I guess I had been thinking of just adding another streamlit based frontend, with the general idea of having one streamlit based module (or set of modules sharing a parent) per "feature set".  More or less just doing more of what you've already done with some of the other parts of PyMedPhys.  And I was anticipating doing the work myself (I'm going to end up going back and forth on implementing pseudonymisation in PyMedPhys and utilising it in another project, OnkoDICOM).
It's a bit. of a bootstrap situation where before the pseudonymisation is available in OnkoDICOM, there is a need for people to pseudonymise their data when they find a problem in OnkoDICOM so that it can be replicated.  Or even when pseudonymisation is available in OnkoDICOM, if the problem occurs before one can pseudonymise...  one needs to step back to PyMedPhys (to avoid whatever the OnkoDICOM issues are).

The above "gui per feature" approach is compatible with an embeddable approach (per windows-embeddable above).  I didn't check to see if there are linux-embeddable and macOS-embeddable equivalents.  If those are available (or will be available), I think that's my preference.  But this results in the developer environment being quite different from the user environment, and possibly requiring "bundling" (not too dissimilar from compiling) for realistic testing.  It's a tradeoff.  I think it depends on how big a user base one expects and the relative value of developer time and effort vs. user time and effort.

Just having an installer (NSIS) doesn't seem to really be enough, but maybe there is more to it that I didn't grasp.

I'm not in favor of using a multiple language approach.  The knowledge required to be productive is very high.
There will also be a language version combinatorics  problem.  It's OK if you have a highly unified organisation with specialisation amongst developers.  The kind of specialisation that is in the opposite direction of what management types tend to think they will get with Agile (I'm not interested in starting an r-war, it's not about the methodology, it is about what management think they get when they adopt it), namely "any team can be assigned to any part of the software" and "any developer can work on any piece of software" (even if they acknowledge there will be some efficiency losses).

The Unix (tm?) approach of small pieces that are tied together using simple interfaces (stdin/stdout) has merit.
Having the PyMedPhys CLI as the basic building block, and GUI separate from that (and separate from the PyMedPhys package itself) invoking the CLI would approximate that.  In which case, you can have people do whatever they like for GUI, but the plumbing stays pure Python. 

I wholly agree with the goal of making things so that beginner developers can be productive contributors.
But that also means not adopting some of the more powerful approaches that are complex in nature.
Some solutions are elegant and sophisticated, or at least very compact, but hard to understand if one doesn't have the (esoteric?) knowledge.

**Comment by SimonBiggs** [comment]
> More or less just doing more of what you've already done with some of the other parts of PyMedPhys

Yup, taking a step back, and letting the other parts of my discussion get kicked down the road "to be solved later". If there was CLI command `pymedphys gui anonymise-dicom` that booted up the DICOM anonymiser with psuedo included, that would keep things simple, and solve what is needed today, today, and what is needed later, later.

I like your initial thought of how you were planning to go about it.

**Comment by SimonBiggs** [comment]
In reality, I was hoping to have the ability to just run `pymedphys gui` and have the GUI open up where the user is greeted by an application index containing all of the PyMedPhys GUIs that have been made.

Users could pin certain applications within the GUI, and also search/filter through the application list.

That way, in the future, any GUI installation/wrapper would just have to call `pymedphys gui` under the hood and present the UI to the user within an native application window (https://superuser.com/a/1421401).

The next major hurdle is actually the following issue:

https://github.com/streamlit/streamlit/issues/543#issuecomment-660410027

Which in a sense is currently being blocked by the following:

https://github.com/streamlit/streamlit/issues/1920#issue-685278392

**Comment by sjswerdloff** [comment]
>In reality, I was hoping to have the ability to just run pymedphys gui and have the GUI open up where the user is greeted by an application index containing all of the PyMedPhys GUIs that have been made.

Yes.  That would be really nice.  Minimal command line to launch the GUI, or on MS Windows a batch file of some sort with an icon link, or a link to an application with parameters.

While streamlit works on https://github.com/streamlit/streamlit/issues/543#issuecomment-660410027 and https://github.com/streamlit/streamlit/issues/1920#issue-685278392

in the meantime... why not/don't we write a streamlit parent app that uses process calls that invoke streamlit for the "applets"?

streamlit pymedphys_parent_gui.py
...

gets a list of gui apps (introspection?)
lists them
...

invokes
streamlit pymedphys gui applet
or 
streamlit my_app.py
or 
something like that.

It's not likely to be efficient or performant, but for "minimal GUI" that is frequently "one and done" that's not necessarily such a big deal.
This is a lot closer to "pick a number" menu stuff back when displays were green and had 80 characters by 24 rows, rather than a graphical diagram editor...

**Comment by SimonBiggs** [comment]
> in the meantime... why not/don't we write a streamlit parent app that uses process calls that invoke streamlit for the "applets"?

That's actually what we've implemented internally. You can see the effect by calling `pymedphys experimental gui` which calls the following function:

https://github.com/pymedphys/pymedphys/blob/4b9fdaec8f0a0488b157a694d8a95ba48cef9ac4/pymedphys/_gui/__init__.py#L50-L71

I am then using nginx to route the different URLs to each port:

https://github.com/CCA-Physics/physics-server/blob/be4cb735b70f663427d23b9b3a95e03aa0c4ec98/RCCC/docker/nginx/nginx.conf#L14-L77

**Comment by SimonBiggs** [comment]
> in the meantime... why not/don't we write a streamlit parent app that uses process calls that invoke streamlit for the "applets"?
> ...
> It's not likely to be efficient or performant, but for "minimal GUI" that is frequently "one and done" that's not necessarily such a big deal.

Also, I'd like to add that given the overwhelming majority of the GUI work would be still going into making streamlit based applications, almost all of the work done would be compatible with the future direction of the PyMedPhys GUI.

Also, it would be very helpful to have something useable and running. Best way to find out how to do it better, in my experience.

So, I say, go for it. I'm happy for you to use the experimental namespace to undergo its namesake. Experiment to your heart's content, and I'm more than happy for experimental parts of pymedphys to be released on pypi, which can then be used in conjunction with OnkoDICOM. The only conditions are that it is booted via `pymedphys experimental something-goes-here` and it doesn't break the testing suite :).

**Comment by SimonBiggs** [comment]
Also, I just stocked up the beer :)

**Comment by sjswerdloff** [comment]
streamlit has a file_uploader widget, which appears to launch the native OS file selection dialog, but it's of the "pick one" variety, and doesn't support selecting a directory.
The forum has mention of a multi-select capability, but it's not implemented yet (from looking at the code).
The suggested alternative is to use something like os.listdir() as input to the multiselect widget.
But that puts the burden on the application/GUI programmer to address navigation.

The reason for using a GUI framework is to avoid having to develop the Widgets.
streamlit seems in decent shape as a web native UI, but I'd rather wait until they get local file selection more complete.
So I think I'd like to put this on the backlog.

**Comment by SimonBiggs** [comment]
Streamlit has a react component builder:
https://docs.streamlit.io/en/stable/develop_streamlit_components.html#react

It allows one to take react components already in the wild and expose them as streamlit components. In the past I have used the following component to allow intuitive file selection within a web application:
https://blueprintjs.com/docs/#core/components/tree

My goals for streamlit are not limited to using its prebaked components. It provides a beautiful framework to expose web components. And I would hope that in the future, some people who install PyMedPhys are actually installing it just to use the Streamlit components we have built for our own use.

Another component that I envisage being very useful for PyMedPhys GUIs is the following:
https://blueprintjs.com/docs/#table/features.editing

**Comment by sjswerdloff** [comment]
>Streamlit components we have built for our own use.
Then we need a place to put these, possibly under the _streamlit directory, or possibly in something like _streamlit_components
It's a detour I'm not sure I'm prepared to take (happy to use what someone else provides, not so sure about putting in the time to create the components)

**Comment by SimonBiggs** [comment]
Yup, all good. Also as an initial GUI users can just be asked to provide a path string as an input?

**Comment by SimonBiggs** [comment]
> streamlit has a file_uploader widget, which appears to launch the native OS file selection dialog, but it's of the "pick one" variety, and doesn't support selecting a directory.
> The forum has mention of a multi-select capability, but it's not implemented yet (from looking at the code).

Looks like it's now implemented:

https://youtu.be/eFKYEqSuy-Q?t=146

**Comment by sjswerdloff** [comment]
I've watched the video, so I'll take a look at the most recent version.
